Container stats query generates exception

## Description
I used the following call to get container stats and got an  exception:
```
    ContainerStats stats = dockerClient.stats(containerId);
```
## How to reproduce
I have certain containers that are in exited status and for some reason for both of these I am getting this error. I don't know how to reproduce it.
## What do you expect
I expect the API code to handle null or null references and not generate any kind of exception.

## What happened instead
I got the following exception instead.

## Software:

- `docker version`: 
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false
- Spotify's docker-client version: 8.9.1

## Full backtrace

```text
com.spotify.docker.client.exceptions.DockerRequestException: Request error: GET unix://localhost:80/containers/2d59dbc285011cc03a6218863ad7832d2586eab3e57cb163273bfeeb5c7b0f3b/stats?stream=0: 200
        at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:2690)
        at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2630)
        at com.spotify.docker.client.DefaultDockerClient.stats(DefaultDockerClient.java:2313)
        .....
Caused by: com.spotify.docker.client.shaded.javax.ws.rs.client.ResponseProcessingException: com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.BlockIoStats, problem: null
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@5d8b3981; line: 1, column: 324] (through reference chain: com.spotify.docker.client.messages.ContainerStats["blkio_stats"])
        at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:806)
        at org.glassfish.jersey.client.JerseyInvocation.access$700(JerseyInvocation.java:92)
        at org.glassfish.jersey.client.JerseyInvocation$5.completed(JerseyInvocation.java:773)
        at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:198)
        at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:79)
        at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:180)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:340)
        at org.glassfish.jersey.client.ClientRuntime$3.run(ClientRuntime.java:210)
        ... 5 more
Caused by: com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.BlockIoStats, problem: null
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@5d8b3981; line: 1, column: 324] (through reference chain: com.spotify.docker.client.messages.ContainerStats["blkio_stats"])
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:277)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1441)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:476)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:495)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:276)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:228)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:138)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:405)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:1583)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:964)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:815)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.invokeReadFrom(ReaderInterceptorExecutor.java:256)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.aroundReadFrom(ReaderInterceptorExecutor.java:235)
        at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor.proceed(ReaderInterceptorExecutor.java:155)
        at org.glassfish.jersey.message.internal.MessageBodyFactory.readFrom(MessageBodyFactory.java:1085)
        at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:874)
        at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:808)
        at org.glassfish.jersey.client.ClientResponse.readEntity(ClientResponse.java:326)
        at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:803)
        ... 17 more
Caused by: java.lang.NullPointerException
        at com.spotify.docker.client.shaded.com.google.common.collect.ImmutableList.copyOf(ImmutableList.java:239)
        at com.spotify.docker.client.messages.BlockIoStats.create(BlockIoStats.java:73)
        at sun.reflect.GeneratedMethodAccessor61.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.introspect.AnnotatedMethod.call(AnnotatedMethod.java:117)
        at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:274)
        ... 40 more
```

--- a/src/main/java/com/spotify/docker/client/messages/BlockIoStats.java
+++ b/src/main/java/com/spotify/docker/client/messages/BlockIoStats.java
@@ -22,43 +22,72 @@ package com.spotify.docker.client.messages;
 
 import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
 import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
-import static com.google.common.collect.ImmutableList.copyOf;
 
 import com.fasterxml.jackson.annotation.JsonAutoDetect;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableList;
-
 import java.util.List;
+import javax.annotation.Nullable;
 
 @AutoValue
 @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
 public abstract class BlockIoStats {
 
+  @Nullable
   @JsonProperty("io_service_bytes_recursive")
   public abstract ImmutableList<Object> ioServiceBytesRecursive();
 
+  @Nullable
   @JsonProperty("io_serviced_recursive")
   public abstract ImmutableList<Object> ioServicedRecursive();
 
+  @Nullable
   @JsonProperty("io_queue_recursive")
   public abstract ImmutableList<Object> ioQueueRecursive();
 
+  @Nullable
   @JsonProperty("io_service_time_recursive")
   public abstract ImmutableList<Object> ioServiceTimeRecursive();
 
+  @Nullable
   @JsonProperty("io_wait_time_recursive")
   public abstract ImmutableList<Object> ioWaitTimeRecursive();
 
+  @Nullable
   @JsonProperty("io_merged_recursive")
   public abstract ImmutableList<Object> ioMergedRecursive();
 
+  @Nullable
   @JsonProperty("io_time_recursive")
   public abstract ImmutableList<Object> ioTimeRecursive();
 
+  @Nullable
   @JsonProperty("sectors_recursive")
   public abstract ImmutableList<Object> sectorsRecursive();
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder ioServiceBytesRecursive(final List<Object> ioServiceBytesRecursive);
+
+    public abstract Builder ioServicedRecursive(final List<Object> ioServicedRecursive);
+
+    public abstract Builder ioQueueRecursive(final List<Object> ioQueueRecursive);
+
+    public abstract Builder ioServiceTimeRecursive(final List<Object> ioServiceTimeRecursive);
+
+    public abstract Builder ioWaitTimeRecursive(final List<Object> ioWaitTimeRecursive);
+
+    public abstract Builder ioMergedRecursive(final List<Object> ioMergedRecursive);
+
+    public abstract Builder ioTimeRecursive(final List<Object> ioTimeRecursive);
+
+    public abstract Builder sectorsRecursive(final List<Object> sectorsRecursive);
+
+    public abstract BlockIoStats build();
+  }
 
   @JsonCreator
   static BlockIoStats create(
@@ -70,9 +99,15 @@ public abstract class BlockIoStats {
       @JsonProperty("io_merged_recursive") final List<Object> ioMergedRecursive,
       @JsonProperty("io_time_recursive") final List<Object> ioTimeRecursive,
       @JsonProperty("sectors_recursive") final List<Object> sectorsRecursive) {
-    return new AutoValue_BlockIoStats(
-        copyOf(ioServiceBytesRecursive), copyOf(ioServicedRecursive), copyOf(ioQueueRecursive),
-        copyOf(ioServiceTimeRecursive), copyOf(ioWaitTimeRecursive), copyOf(ioMergedRecursive),
-        copyOf(ioTimeRecursive), copyOf(sectorsRecursive));
+    return new AutoValue_BlockIoStats.Builder()
+        .ioServiceBytesRecursive(ioServiceBytesRecursive)
+        .ioServicedRecursive(ioServicedRecursive)
+        .ioQueueRecursive(ioQueueRecursive)
+        .ioServiceTimeRecursive(ioServiceTimeRecursive)
+        .ioWaitTimeRecursive(ioWaitTimeRecursive)
+        .ioMergedRecursive(ioMergedRecursive)
+        .ioTimeRecursive(ioTimeRecursive)
+        .sectorsRecursive(sectorsRecursive)
+        .build();
   }
 }

--- a/src/main/java/com/spotify/docker/client/messages/CpuStats.java
+++ b/src/main/java/com/spotify/docker/client/messages/CpuStats.java
@@ -29,6 +29,7 @@ import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableList;
 import java.util.List;
+import javax.annotation.Nullable;
 
 @AutoValue
 @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
@@ -37,18 +38,35 @@ public abstract class CpuStats {
   @JsonProperty("cpu_usage")
   public abstract CpuUsage cpuUsage();
 
+  @Nullable
   @JsonProperty("system_cpu_usage")
   public abstract Long systemCpuUsage();
 
   @JsonProperty("throttling_data")
   public abstract ThrottlingData throttlingData();
 
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder cpuUsage(final CpuUsage cpuUsage);
+
+    public abstract Builder systemCpuUsage(final Long systemCpuUsage);
+
+    public abstract Builder throttlingData(final ThrottlingData throttlingData);
+
+    public abstract CpuStats build();
+  }
+
   @JsonCreator
   static CpuStats create(
       @JsonProperty("cpu_usage") final CpuUsage cpuUsage,
       @JsonProperty("system_cpu_usage") final Long systemCpuUsage,
       @JsonProperty("throttling_data") final ThrottlingData throttlingData) {
-    return new AutoValue_CpuStats(cpuUsage, systemCpuUsage, throttlingData);
+    return new AutoValue_CpuStats.Builder()
+        .cpuUsage(cpuUsage)
+        .systemCpuUsage(systemCpuUsage)
+        .throttlingData(throttlingData)
+        .build();
   }
 
   @AutoValue
@@ -57,6 +75,7 @@ public abstract class CpuStats {
     @JsonProperty("total_usage")
     public abstract Long totalUsage();
 
+    @Nullable
     @JsonProperty("percpu_usage")
     public abstract ImmutableList<Long> percpuUsage();
 
@@ -66,14 +85,32 @@ public abstract class CpuStats {
     @JsonProperty("usage_in_usermode")
     public abstract Long usageInUsermode();
 
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      public abstract Builder totalUsage(final Long totalUsage);
+
+      public abstract Builder percpuUsage(final List<Long> percpuUsage);
+
+      public abstract Builder usageInKernelmode(final Long usageInKernelmode);
+
+      public abstract Builder usageInUsermode(final Long usageInUsermode);
+
+      public abstract CpuUsage build();
+    }
+
     @JsonCreator
     static CpuUsage create(
         @JsonProperty("total_usage") final Long totalUsage,
         @JsonProperty("percpu_usage") final List<Long> perCpuUsage,
         @JsonProperty("usage_in_kernelmode") final Long usageInKernelmode,
         @JsonProperty("usage_in_usermode") final Long usageInUsermode) {
-      return new AutoValue_CpuStats_CpuUsage(totalUsage, ImmutableList.copyOf(perCpuUsage),
-          usageInKernelmode, usageInUsermode);
+      return new AutoValue_CpuStats_CpuUsage.Builder()
+          .totalUsage(totalUsage)
+          .percpuUsage(perCpuUsage)
+          .usageInKernelmode(usageInKernelmode)
+          .usageInUsermode(usageInUsermode)
+          .build();
     }
   }
 

--- a/src/test/java/com/spotify/docker/client/messages/ContainerStatsTest.java
+++ b/src/test/java/com/spotify/docker/client/messages/ContainerStatsTest.java
@@ -46,4 +46,16 @@ public class ContainerStatsTest {
     objectMapper.readValue(fixture("fixtures/1.30/containerStats.json"), ContainerStats.class);
   }
 
+  @Test
+  public void test1_33_running() throws Exception {
+    objectMapper.readValue(fixture(
+        "fixtures/1.33/containerStatsRunning.json"), ContainerStats.class);
+  }
+
+  @Test
+  public void test1_33_exited() throws Exception {
+    objectMapper.readValue(fixture(
+        "fixtures/1.33/containerStatsExited.json"), ContainerStats.class);
+  }
+
 }

--- a/src/test/resources/fixtures/1.33/containerStatsExited.json
+++ b/src/test/resources/fixtures/1.33/containerStatsExited.json
@@ -1,0 +1,44 @@
+{
+  "blkio_stats": {
+    "io_merged_recursive": null,
+    "io_queue_recursive": null,
+    "io_service_bytes_recursive": null,
+    "io_service_time_recursive": null,
+    "io_serviced_recursive": null,
+    "io_time_recursive": null,
+    "io_wait_time_recursive": null,
+    "sectors_recursive": null
+  },
+  "cpu_stats": {
+    "cpu_usage": {
+      "total_usage": 0,
+      "usage_in_kernelmode": 0,
+      "usage_in_usermode": 0
+    },
+    "throttling_data": {
+      "periods": 0,
+      "throttled_periods": 0,
+      "throttled_time": 0
+    }
+  },
+  "id": "3c6271139fb2c54e3f2eadf3c7fc9a95887e3d47f24a0caef615b28f111ccfa3",
+  "memory_stats": {},
+  "name": "/jovial_liskov",
+  "num_procs": 0,
+  "pids_stats": {},
+  "precpu_stats": {
+    "cpu_usage": {
+      "total_usage": 0,
+      "usage_in_kernelmode": 0,
+      "usage_in_usermode": 0
+    },
+    "throttling_data": {
+      "periods": 0,
+      "throttled_periods": 0,
+      "throttled_time": 0
+    }
+  },
+  "preread": "0001-01-01T00:00:00Z",
+  "read": "0001-01-01T00:00:00Z",
+  "storage_stats": {}
+}

--- a/src/test/resources/fixtures/1.33/containerStatsRunning.json
+++ b/src/test/resources/fixtures/1.33/containerStatsRunning.json
@@ -1,0 +1,172 @@
+{
+  "blkio_stats": {
+    "io_merged_recursive": [],
+    "io_queue_recursive": [],
+    "io_service_bytes_recursive": [
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Read",
+        "value": 7155712
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Write",
+        "value": 0
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Sync",
+        "value": 0
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Async",
+        "value": 7155712
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Total",
+        "value": 7155712
+      }
+    ],
+    "io_service_time_recursive": [],
+    "io_serviced_recursive": [
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Read",
+        "value": 284
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Write",
+        "value": 0
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Sync",
+        "value": 0
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Async",
+        "value": 284
+      },
+      {
+        "major": 8,
+        "minor": 0,
+        "op": "Total",
+        "value": 284
+      }
+    ],
+    "io_time_recursive": [],
+    "io_wait_time_recursive": [],
+    "sectors_recursive": []
+  },
+  "cpu_stats": {
+    "cpu_usage": {
+      "percpu_usage": [
+        21169594,
+        18112083
+      ],
+      "total_usage": 39281677,
+      "usage_in_kernelmode": 10000000,
+      "usage_in_usermode": 20000000
+    },
+    "online_cpus": 2,
+    "system_cpu_usage": 305318850000000,
+    "throttling_data": {
+      "periods": 0,
+      "throttled_periods": 0,
+      "throttled_time": 0
+    }
+  },
+  "id": "624a23cb3a58da0fd54b3a9dc1cd9a5f545240b2fd92549a5abb422700b401ac",
+  "memory_stats": {
+    "limit": 2099351552,
+    "max_usage": 8720384,
+    "stats": {
+      "active_anon": 1425408,
+      "active_file": 823296,
+      "cache": 7200768,
+      "dirty": 4096,
+      "hierarchical_memory_limit": 9223372036854771712,
+      "hierarchical_memsw_limit": 9223372036854771712,
+      "inactive_anon": 8192,
+      "inactive_file": 6340608,
+      "mapped_file": 4898816,
+      "pgfault": 1166,
+      "pgmajfault": 49,
+      "pgpgin": 2548,
+      "pgpgout": 449,
+      "rss": 1396736,
+      "rss_huge": 0,
+      "swap": 0,
+      "total_active_anon": 1425408,
+      "total_active_file": 823296,
+      "total_cache": 7200768,
+      "total_dirty": 4096,
+      "total_inactive_anon": 8192,
+      "total_inactive_file": 6340608,
+      "total_mapped_file": 4898816,
+      "total_pgfault": 1166,
+      "total_pgmajfault": 49,
+      "total_pgpgin": 2548,
+      "total_pgpgout": 449,
+      "total_rss": 1396736,
+      "total_rss_huge": 0,
+      "total_swap": 0,
+      "total_unevictable": 0,
+      "total_writeback": 0,
+      "unevictable": 0,
+      "writeback": 0
+    },
+    "usage": 8720384
+  },
+  "name": "/inspiring_raman",
+  "networks": {
+    "eth0": {
+      "rx_bytes": 828,
+      "rx_dropped": 0,
+      "rx_errors": 0,
+      "rx_packets": 10,
+      "tx_bytes": 0,
+      "tx_dropped": 0,
+      "tx_errors": 0,
+      "tx_packets": 0
+    }
+  },
+  "num_procs": 0,
+  "pids_stats": {
+    "current": 2
+  },
+  "precpu_stats": {
+    "cpu_usage": {
+      "percpu_usage": [
+        21169594,
+        18112083
+      ],
+      "total_usage": 39281677,
+      "usage_in_kernelmode": 10000000,
+      "usage_in_usermode": 20000000
+    },
+    "online_cpus": 2,
+    "system_cpu_usage": 305316870000000,
+    "throttling_data": {
+      "periods": 0,
+      "throttled_periods": 0,
+      "throttled_time": 0
+    }
+  },
+  "preread": "2017-11-22T04:04:20.31756147Z",
+  "read": "2017-11-22T04:04:21.318496884Z",
+  "storage_stats": {}
+}

